MaeBlock.cpp

```diff
@@ -44,53 +191,186 @@ bool BufferedIndexedBlockMap::hasIndexedBlock(const std::string& name)
+        unsigned int index)
+{
+    for(const auto& p : properties) {
+        const auto& property = p.second;
+        if(property->isDefined(index)) {
+            out << " " << local_to_string(property->at(index));
+        } else {
+            out << " <>";
+        }
+    }
+}
+
+void IndexedBlock::write(ostream& out, unsigned int current_indentation) const
+{
+    string root_indentation = string(current_indentation, ' ');
+    string indentation = string(current_indentation+2, ' ');
+    const bool has_data = m_bmap.size() + m_rmap.size() + m_imap.size() + m_smap.size() > 0;
+
+    out << root_indentation << getName() << "[" << to_string((int)size()) << "] {\n";
+
+    if(has_data) {
+        out << indentation + "# First column is Index #\n";
+    }
+
+    output_property_names(out, indentation, m_bmap);
+    output_property_names(out, indentation, m_rmap);
+    output_property_names(out, indentation, m_imap);
+    output_property_names(out, indentation, m_smap);
+
+    if(has_data) {
+        out << indentation + ":::\n";
+    }
+
+    for(unsigned int i=0; i<size(); i++) {
+        out << indentation << i+1;
+        output_indexed_property_values(out, indentation, m_bmap, i);
+        output_indexed_property_values(out, indentation, m_rmap, i);
+        output_indexed_property_values(out, indentation, m_imap, i);
+        output_indexed_property_values(out, indentation, m_smap, i);
+        out << endl;
+    }
+
+    if(has_data) {
+        out << indentation + ":::\n";
+    }
+
+    out << root_indentation << "}\n";
+
+    return;
+}
+
+string IndexedBlock::toString() const
+{
+    ostringstream stream;
+    write(stream);
+
+    return stream.str();
+}
+
+template <typename T>
+bool IndexedProperty<T>::operator==(const IndexedProperty<T>& rhs) const
+{
+    if(m_is_null == nullptr || rhs.m_is_null == nullptr) {
+        if((m_is_null == nullptr) != (rhs.m_is_null == nullptr)) return false;
+    } else if(*m_is_null != *(rhs.m_is_null)) {
+        return false;
+    }
+    if(m_data != rhs.m_data) return false;
+    return true;
+}
+
+
+// For doubles we need to implement our own comparator for the vectors to
+// take precision into account
+template <>
+bool IndexedProperty<double>::operator==(const IndexedProperty<double>& rhs) const
 {
-    set_indexed_property<IndexedProperty<std::string>>(m_smap, name, value);
+    if(m_is_null == nullptr || rhs.m_is_null == nullptr) {
+        if((m_is_null == nullptr) != (rhs.m_is_null == nullptr)) return false;
+    } else if(*m_is_null != *(rhs.m_is_null)) {
+        return false;
+    }
+    for(int i=0; i<m_data.size(); i++)
+        if(abs(m_data[i] - rhs.m_data[i]) > tolerance) return false;
+
+    return true;
+}
+
+bool IndexedBlock::operator==(const IndexedBlock& rhs) const
+{
+    using namespace std;
+
+    for(const auto& p : m_bmap) {
+        if(rhs.m_bmap.count(p.first) != 1) return false;
+        const auto& bmap2 = rhs.m_bmap.find(p.first);
```

Comment:
This doesn't return `false` if rhs has more elements.

Also, it does a duplicate find in the map (here and the other checks). Since these are ordered maps, and you're checking whether all the elements match, maybe something like:
   
    if (m_bmap.size() != rhs.m_bmap.size()) return false;
    auto diff = std::mismatch(m_bmap.begin(), m_bmap.end(), rhs.m_bmap.begin(),
     [](const shared_ptr<IndexedProperty<double>>& l, const shared_ptr<IndexedProperty<double>>& r) {return l.first == r.first && *(l.second) == *(r.second)});
    if (diff.first != m_bmap.end()) return false;

I suppose that the lambda that compares pointer values could be a template for use across this function.